best-practices/grpc-apis.md

```diff
@@ -0,0 +1,105 @@
+
+Maintaining downgrade compatibility when launching a new release requires that
+the release not include any features not supported by the previous release. In
+the strictest sense this means that new features cannot be "enabled" until one
+release after they are deemed "working". As a concrete example, when Google
+Compute Engine introduced multi-queue networking we did so one release later
+than when it was "fully supported" in GCE production. This allowed us to safely
+roll-back the release when a bug elsewhere was encountered while not bricking
+VMs that had booted after multi-queue had been enabled.
+
+TODO: Find a concrete example from Kata rather than from an external source.
+
+### R3. Enums are Problematic
+
+Related to R2, enum types are problematic as adding additional values in a new
+releases can create undefined behavior in prior releases.
+
+### R4. Rarely Change Field Types
+
+Related to the R1, while not always strictly unsafe it is usually unwise to
+change the type of an existing field. In most cases it is better to inroduce a
+new field with the new type and slowly phase out the old field once it has left
+the backward compatibility window.
+
+Exceptions to this rule can include converting enum types to int32 or int64,
+provided no values other than the original enum definition are used
+
+## Guidelines
+
+### GL1. Have Clearly Defined Compatibility Windows
+
+TODO: Change this guideline to encompass Kata's agreed-upon compatibiltiy
+policy.
+
+It is critical to have a clear policy on what upgrade and downgrade paths are
+supported. One policy that has been generally successful for other large
+projects (Google Compute Engine, in particular) it to support unbounded forward
+upgrades and single-release downgrades.
+
+### GL2. The Scope of Compatibility is Only Releases that may Mix
+
+Many compatibility problems can been addressed be avoiding mixing components.
+For example, if each Kata release includes both the agent and the runtime, the
+agent can be
+
+### GL3. Redundancy is a Useful Tool
+
+While it can be aesthetically displeasing, redundantly encoding similar (or
+identical) information can allow slow migration from one behavior
+
+### GL4. Changes to Defaults Should Straddle Compatibility Boundaries
+
+As per R2 above, new fields cannot generally be introduced and used in the same
+release. More broadly, new behaviors (typically new features) should not become
+the default until the _oldest_ release not supporting that behavior is no longer
+in the downgrade compatibiltiy window.
+
+### GL5. On-the-Wire Versioning is Rarely Useful
+
+Very few protocol buffers at Google include version information as part of the
+wire format (either explicitly or implicitly). For example, the Google Compute
+Engine API is still on "version 1" despite dozens of calls and arguments being
+added since its inception. This is possible because new features are rolled out
+as "additions" to existing APIs, and by following R2 above.
```

Comment:
As kata-agent will keeps running after kata-runtime is upgraded, I think we need a mechanism to let   kata-runtime know which field isn't supported by this version of kata-agent, and kata-runtime should downgrade it's grpc message version to appropriate version for better adapt to the agent.  so question is, without on wire versioning support, how could kata-runtime do this？

Reply:
One way I think is, we give the on-disk storage config of a POD a version number, and this version number will represent the kata-agent version, kata-runtime will send appropriate message according to the on-disk version(which equals to agent version). What do you think, do you have better practice to share?

Reply:
I think that @jon is suggesting we should not perform any version checks.

A possible solution which was mentioned briefly yesterday is that we could introduce a new `RestartAgent` gRPC service. When we can guarantee that all agent versions support that service (and assuming that adding new services to the protocol will not stop `RestartAgent` from working), when the runtime is upgraded, it can call `RestartService` for running pods. That will cause the agent to re-exec itself, reloading a newer binary version (which supports the newer gRPC protocol version). Once done, the runtime is guaranteed that it can use the full (new) gRPC protocol version when talking to the agent. This would potentially require the agent to pass state between the original instance and the newly-exec'd instance but that should be achievable.

Alternatively, rather than re-execing the agent, we could apparently squirt bytes down the wire via gRPC from the runtime to the agent so that the agent could read the byte stream and create a new agent instance out of it (write the bytes to disk and re-exec?). I think @jon mentioned that would be slower though?

I'd vote for the re-exec option as it's proven and used by init systems such as systemd and upstart.

One issue with both approaches is being able to rewrite the agent binary. This might not be possible or desirable. We could share the agent into all pods images via 9p and then the runtime would only need to update the (host) agent binary file and ask all the running agent instances to re-exec to use it. Again, performance might take a hit here due to 9p.

Another concern is of course security - arbitrarily rewriting files isn't ideal. We'd need to think carefully about checksums, security tokens, gpg signatures, etc.

Reply:
How long it takes to get a new agent binary into the VM, and how, I'm not so worried about. This will be a very rare sequence (it's not like something we would be doing every minute, or even daily - this is probably a once every 3 month or more type scenario I would think).
How we hand over the state from the old agent to the new I think will be an interesting challenge - but, as you say, should be achievable.

Reply:
For Upstart, we used a pipe - see https://wiki.ubuntu.com/FoundationsTeam/Specs/QuantalUpstartStatefulReexec#State-Passing for the logic.